### DISCUSSION TITLE
Create `extendable` object

### DISCUSSION THREAD

**Discussion by roninjin10** [wevm/viem#2239] [Idea / Feature Request]
Viem uses a decorator pattern to add functionality to a base client using `extend`. It's a nice pattern that [tevm](https://tevm.sh) follows for both it's clients and it's lower level primitive to allow better code splitting. [One of many examples](https://github.com/evmts/tevm-monorepo/blob/main/packages/base-client/src/BaseClient.ts#L73)

Idea here is to seperate `extendable` into it's own lower level primitive that Tevm and anyone else can reuse. I might even suggest making transports and some other objects in Viem extendable.

```typescript
export const createExtendable = <TObject extends {}>(obj: TObject): Extandable<TObject>
  function extend(base: typeof obj) {
    type ExtendFn = (base: typeof obj) => unknown
    return (extendFn: ExtendFn) => {
      const extended = extendFn(base) as Extended
      for (const key in obj) delete extended[key]
      const combined = { ...base, ...extended }
      return Object.assign(combined, { extend: extend(combined as any) })
    }
  }

  return Object.assign(obj, { extend: extend(obj) as any })
}
```

**Comment by tmm** [comment]
Interesting idea! Not sure this pattern scales very well though so would need to think about if we want to add it to additional types. Also, not sure it's worth exporting this API from Viem as it doesn't really mesh with the surface area that Viem's public API covers.

**Comment by roninjin10** [comment]
> Also, not sure it's worth exporting this API from Viem as it doesn't really mesh with the surface area that Viem's public API covers.

I agree with this it would actually make more sense for extendable to be it's own package I think

**Reply by tmm** [reply]
```
npm i extendable
```

**Comment by roninjin10** [comment]
ok i'll make this in tevm repo and circle back later to see if sharing code makes sense
